src/cmd/support_archive/builder.go

```diff
@@ -0,0 +1,95 @@
+package support_archive
+
+import (
+	"context"
+	"os"
+
+	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/src/api/v1beta1"
+	"github.com/Dynatrace/dynatrace-operator/src/scheme"
+	"github.com/Dynatrace/dynatrace-operator/src/version"
+	"github.com/spf13/cobra"
+)
+
+const (
+	use               = "support-archive"
+	namespaceFlagName = "namespace"
+	stdoutFlagName    = "stdout"
+)
+
+var (
+	namespaceFlagValue string
+	stdoutFlagValue    bool
+)
+
+type CommandBuilder struct {
+}
+
+func NewCommandBuilder() CommandBuilder {
+	return CommandBuilder{}
+}
+
+func (builder CommandBuilder) Build() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:  use,
+		Long: "Pack logs and manifests useful for troubleshooting into single tarball",
+		RunE: builder.buildRun(),
+	}
+	addFlags(cmd)
+	return cmd
+}
+
+func addFlags(cmd *cobra.Command) {
+	cmd.PersistentFlags().StringVar(&namespaceFlagValue, namespaceFlagName, "dynatrace", "Specify a different Namespace.")
+	cmd.PersistentFlags().BoolVar(&stdoutFlagValue, stdoutFlagName, false, "Write tarball to stdout.")
+}
+
+func (builder CommandBuilder) buildRun() func(*cobra.Command, []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		ctx := supportArchiveContext{
+			ctx:           context.TODO(),
+			namespaceName: namespaceFlagValue,
+			toStdout:      stdoutFlagValue,
+		}
+
+		registerSupportArchiveLogger(&ctx)
+		version.LogVersionToLogger(ctx.log)
+
+		err := dynatracev1beta1.AddToScheme(scheme.Scheme)
+		if err != nil {
+			return err
+		}
+
+		supportArchive, err := newTarball(&ctx)
+		if err != nil {
+			return err
+		}
+		defer supportArchive.close()
+
+		runCollectors(&ctx, supportArchive)
+		printCopyCommand(&ctx, supportArchive)
+
+		return nil
+	}
+}
+
+func runCollectors(ctx *supportArchiveContext, supportArchive *tarball) {
+	collectors := []func(*supportArchiveContext, *tarball) error{
+		collectOperatorVersion,
+	}
+
+	for _, c := range collectors {
+		if err := c(ctx, supportArchive); err != nil {
+			logErrorf(ctx.log, err, "failed collector")
+		}
+	}
+}
+
+func printCopyCommand(ctx *supportArchiveContext, supportArchive *tarball) {
```

Comment:
why not a method of `supportArchiveContext` ?

then you could "mock" the `os.Getenv` to some `ctx.GetPodNamespace` and `ctx.GetPodName` or something so it can be tested

Reply:
Again, printing this command isn't a task of the context, it just holds information. You can't put everything that uses the context into the context, because then you end up with a big mess of unrelated functions. That's the nature of a context, that it's used in different occasions, that doesn't imply it has to own all the stuff it is used for.

Reply:
passing around a big struct that has a name that barley means anything into simple function can be reeally bad, because if the simple function uses 1 (maybe 2) fields out of the 6 fields of the struct then 
- its annoying to test (which 2 fields is needed again out of the 6 so I don't get a nil pointer?) 
- horrible to read the code (what is the function actually using and doing with the struct? this is further worsened by the fact you pass in the struct every where as a reference so anything can modify it which is misleading at best and dangerous at worst )

Reply:
You're right about the pointers, will check that.
The one thing about the struct is also, that all collector functions (we know there will be a few more in the future) have a common signature. Each of those are then using what they need from the context (there will be a few more fields like API reader, http client, ...). So we have a few options:
* pass all fields to all collector functions. That's equally confusing because not all functions will use all fields AND it will break the 3 param max rule, e.g.
```
type collectorFunc func(ctx context.Context, log logr.Logger, apiReader client.Reader, clientSet kubernetes.Interface, namespace string, tb *tarball)
```
* don't use a common signature, which will de-loop `runCollectors` and quadruple the error handling code there.
```
err := collectOperatorVersion(log, tarball)
if err != nil {
   logErrorf(ctx.log, err, "failed collector")
}

err = collectOperatorLogs(ctx, log, clientSet, namespace, tarball)
if err != nil {
   logErrorf(ctx.log, err, "failed collector")
}

err = collectK8SObjects(ctx, log, apiReader, tarball)
if err != nil {
   logErrorf(ctx.log, err, "failed collector")
}

...
```



So for the collector functions, I think the context is still the best solution.